library-best-practices/glossary.md

```diff
@@ -0,0 +1,84 @@
+Java Dependency Glossary
+------------------------
+
+### Types of conflicts and compatibility
+
+- **Linkage conflict**: The signature of a non-private method or class in a
```

Comment:
It occurs to me we can define linkage conflicts without reference to compile time. I need to think about this more, but something like:

A class file references another method whose signature is not present in the classpath. Linkage conflicts typically manifest because the signature of a non-private method or class has changed in an incompatible way between compile time and runtime.

Reply:
I think the compile-time part helps people form a mental model in a concrete way, and I think it's worth keeping for the definition of "linkage conflict". Also, the conflicts are not limited to the absence of signatures; one case noted is the switch of a class to final, such that derived classes are not possible any more.

In addition to "linkage conflict," I think it's worth defining a more general (superordinate) idea of "linkage error," which would both linkage conflicts and completely missing references (which can be missing dependencies). WDYT?

Reply:
https://docs.oracle.com/javase/specs/jls/se8/jls8.pdf

Looks like I have some light reading to do... 788 pages of it (for Java 8).

Reply:
I will add the non-signature stuff that needs to be mentioned, but I'm not sure what you have in mind for discussing byte code.